ios_mcp/source/SubmitSystemData.c

```diff
@@ -0,0 +1,415 @@
+    memcpy(pd->system_model, &seeprom[0xB8], sizeof(pd->system_model));
+    memcpy(pd->mfg_date, &seeprom[0xC4], sizeof(pd->mfg_date));
+    memcpy(&pd->sec_level, &otp[0x080], sizeof(pd->sec_level));
+    pd->boardType = seeprom[0x21];
+    pd->boardRevision = seeprom[0x22];
+    pd->bootSource = seeprom[0x23];
+    pd->ddr3Size = seeprom[0x24];
+    pd->ddr3Speed = seeprom[0x25];
+    pd->sataDevice = seeprom[0x2C];
+    pd->consoleType = seeprom[0x2D];
+
+    // BSP revision
+    int res = bspGetHardwareVersion(&pd->bsp_rev);
+    if (res != 0) {
+        pd->bsp_rev = 0;
+    }
+
+    // System serial number
+    // NOTE: Assuming code+serial doesn't exceed 15 chars (plus NULL).
+    snprintf(pd->system_serial, sizeof(pd->system_serial), "%s%s",
+        (const char*)&seeprom[0xAC], (const char*)&seeprom[0xB0]);
+    // Mask the last 3 digits of the system serial number.
+    for (unsigned int i = sizeof(pd->system_serial)-1; i > 3; i--) {
+        if (pd->system_serial[i] <= 0x20) {
+            pd->system_serial[i] = 0;
+            continue;
+        }
+        // Found printable text.
+        // Mask the last three digits.
+        pd->system_serial[i-0] = '*';
+        pd->system_serial[i-1] = '*';
+        pd->system_serial[i-2] = '*';
+        break;
+    }
+
+    // Region information
+    int productArea_id, gameRegion;
+    res = getRegionInfo(&productArea_id, &gameRegion);
+    if (res == 0) {
+        // Region info obtained. (If not obtained, these will be 0.)
+        pd->productArea = (uint8_t)productArea_id;
+        pd->gameRegion = (uint8_t)gameRegion;
+    }
+
+    // MLC information
+    MDReadInfo();	// TODO: Check for errors?
+    for (unsigned int i = 0; i < 2; i++) {
+        // We want the first MLC device.
+        MDBlkDrv* drv = MDGetBlkDrv(i);
+        if (!drv->registered || drv->params.device_type != SAL_DEVICE_TYPE_MLC) {
+            continue;
+        }
+
+        pd->mlc.mid_prv = drv->params.mid_prv;
+        pd->mlc.numBlocks = drv->params.numBlocks;
+        pd->mlc.blockSize = drv->params.blockSize;
+        memcpy(pd->mlc.name1, drv->params.name1, sizeof(pd->mlc.name1));
+        break;
+    }    
+
+    // Hash the OTP ROM
+    // TODO: Check for errors.
+    uint8_t hash_ctx[IOSC_HASH_CONTEXT_SIZE];
+    res = IOSC_GenerateHash(hash_ctx, sizeof(hash_ctx), NULL, 0, IOSC_HASH_FLAGS_INIT, NULL, 0);
+    if (res == 0) {
+        res = IOSC_GenerateHash(hash_ctx, sizeof(hash_ctx), otp, 0x400, IOSC_HASH_FLAGS_FINALIZE, pd->otp_sha256, sizeof(pd->otp_sha256));
+    }
+
+    // Hash the post data.
+    // TODO: Check for errors.
+    res = IOSC_GenerateHash(hash_ctx, sizeof(hash_ctx), NULL, 0, IOSC_HASH_FLAGS_INIT, NULL, 0);
+    if (res == 0) {
+        res = IOSC_GenerateHash(hash_ctx, sizeof(hash_ctx), (uint8_t*)pd, sizeof(*pd), IOSC_HASH_FLAGS_UPDATE, NULL, 0);
+        if (res == 0) {
+            // Add some more stuff to the hash.
+            // NOTE: Reusing the OTP buffer here.
```

Comment:
Can you make `dataBuffer` a `uint8_t*` and use that everywhere? Using a buffer called `otp` everywhere can be a bit misleading.

Reply:
See above.